### DISCUSSION TITLE
Distributed Parallel Optimization

### DISCUSSION THREAD

**Discussion by Vejahatf** [snl-dakota/dakota#48] [Q&A] [answered]
I have been using Dakota to perform multi objective function  optimization on a commercial software package, modeling pipeline hydraulics. I have written a .NET driver (.exe) which automatically parses  param files, runs the simulation and writes out the objective functions to results file . The commercial package is single-threaded and no parallelization in the code is possible.
 So far I have successfully used local concurrency on my workstations with no issue and everything works as expected. However, due to the nature of the simulation runs which take up to a few hours to run and the sheer number of optimization runs needed to find the optimum solution I have been thinking to use other non-queued workstations distributed across the same network and perhaps use more core to expedite the whole process. I am mostly using derivative free optimization techniques such MOGA, SOGA, coliny-ea,...
I have been reading through the manual and tried to use Mocrosfot-MPI, mpiexece  to distribute evaluations across all cores available on all machines. However, I have not been able to figure out how to parallelize iterators. I am using hybrid model specifying 8 cores on each machines (async evaluation) and about 8 machines on the network passed to mpiexec machinefile. 
When I submit the job it seems each machine starts running independent optimization runs and not sure what I am missing. 
So my question is whether there are any documents or examples to elaborate on the steps required to parallelize iterators are machines on the same network? 
Dakota and commercial package are installed on all machines and properly setup.
Some details such as such as how to setup working directories for each machine? sharing(copying) the model on each machine or use a UNC path, etc. are not very clear. 
I am mainly working with Windows Server 2016 OS. 
Any help would be highly appreciated.

**Comment by jadamstephens** [comment]
Our downloads of Dakota for Windows are not built with parallel support. That's the reason you're seeing a bunch of independent optimization runs: the mpi launcher is just launching a bunch of serial Dakota instances.

I suggest that you create a driver that can take advantage of the available computing resources. I don't know anything about .NET or Microsoft-MPI, so I can't give very specific recommendations. I do have some experience setting up the kind of workflow you're describing on linux-based HPCs, though.

If Microsoft-MPI works using machine files, you might try setting up Dakota and your driver to do something like this:

* Use Dakota's [batch](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/interface-batch.html) interface. This causes Dakota to write all available parameter sets to one parameters file, launch one driver instance, and expect back a single results file for all the sets. This allows your driver to know about all the work that needs to be done so that it can easily assign it to available computing resources.
* In your driver, you'll have to track which cores on which machines are available, and also which parameter sets need to be run. Initially launch N*8 simulations, where N is the number of available machines, using  mpiexec and the appropriate machine file. As simulations complete, backfill with the next one that hasn't been run. Repeat until all the simulations in the batch have been completed.

**Reply by Vejahatf** [reply]
thanks for the reply, certainly good suggestion I can build my driver with batch interface and Akka.net to perform distributed computation. 
Having said that, I was Just wondering if it is possible to compile/build windows artifacts with parallel MPI support? I have not looked at the source code but if the source code assuming C++ includes MPI header and MPI function calls for parallel processing it should be possible to compile the source code using MS-MPI. Any Thought on this?

**Reply by jadamstephens** [reply]
We've never tried to build Dakota on Windows with MS-MPI (or any MPI), so I don't know what kind of problems you might encounter. I don't know of any reason it shouldn't work.

**Comment by rafmag** [comment]
I have a very similar issue but I'm actually running on linux, the CEE compute clusters. I have an optimization problem that runs just fine in serial, but when I try to launch Dakota using `mpirun -np 50 dakota .....` It's still running as `MPI Dakota Serial`. Any suggestions?

**Comment by Averis007** [comment, marked as answer]
@Vejahatf I'm trying to do something similar. Trying to run dakota on a mini Windows cluster where dakota will ideally run on the master node and submit the job on N*M cores (N computes with M core each). Trying to run jobs through both MS MPI and Intel MPI didn't yield much success. If you made any progress, please share how you went on to implement this.

**Reply by Vejahatf** [reply]
Unfortunately windows version of Dakota does not support MPI. So I had to find another solution for myself and indeed I found an even better solution

My solution is to Use Dakota with Jenkins. I setup Jenkins with multiple nodes for parallel execution of studies.
I wrote a python script for Dakota which remotely triggered jobs to Jenkins and Jenkins returned result.out file back to Dakota at the end of each job build. This is much simpler than modifying Dakota to support MS MPI in windows environment and you have much more control on computational nodes and individual jobs on Jenkins interface. 
This solved my problem.
